Dockerfile

```diff
@@ -0,0 +1,41 @@
+FROM golang:1.18-alpine3.16 as builder
+
+RUN apk add --no-cache \
+    make \
+    git \
+    bash \
+    curl \
+    findutils \
+    gcc \
+    libc-dev \
+    nodejs \
+    npm \
+    openjdk11
+
+RUN npm install -g pnpm
```

Comment:
I have tried, it's very slow to build the frontend project inside the docker, and we can't use the `node_modules` cache. I recommend to build it outside the docker, and just copy the generated `dist` folder to the docker, this can also reduce the modification for the Makefile.

you can refer the clinic-ui repo github action.

Reply:
IMHO, we should make dockerfile self-consistent and having no outside dependencies.
Based on this principle, I've done two things to use docker layer cache with github action for speeding up building process:

- Cache the npm module and go module to different docker image layser https://github.com/pingcap/tidb-dashboard/pull/1439/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R25
- Use the ability of docker action to integrate layer cache to github action cache https://github.com/pingcap/tidb-dashboard/pull/1439/files#diff-454793a6142a49e20ff934e6dfd439fd81e37d3902a05bbe4f18397e26d95757R38

After those, docker building time cost shoud be reduced reasonably.